symengine/parser/sbml/sbml_parser.cpp

```diff
@@ -0,0 +1,345 @@
+#include <algorithm>
+#include <cctype>
+
+using SymEngine::ceiling;
+using SymEngine::floor;
+using SymEngine::gamma;
+using SymEngine::logical_xor;
+using SymEngine::Lt;
+using SymEngine::piecewise;
+using SymEngine::pow;
+using SymEngine::zero;
+
+namespace SymEngine
+{
+
+RCP<const Basic>
+parse_sbml(const std::string &s, bool convert_xor,
+           const std::map<const std::string, const RCP<const Basic>> &constants)
+{
+    SbmlParser p(constants);
+    return p.parse(s, convert_xor);
+}
+
+RCP<const Basic> SbmlParser::parse(const std::string &input, bool convert_xor)
+{
+    inp = input;
+    if (convert_xor) {
+        std::replace(inp.begin(), inp.end(), '^', '@');
+    }
+    m_tokenizer.set_string(inp);
+    if (sbmlparse(*this) == 0)
+        return this->res;
+    throw ParseError("Parsing Unsuccessful");
+}
+
+static std::string lowercase(const std::string &str)
+{
+    std::string lower = str;
+    std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);
+    return lower;
+}
+
+// reference :
+// http://stackoverflow.com/questions/30393285/stdfunction-fails-to-distinguish-overloaded-functions
+typedef RCP<const Boolean> (*double_arg_boolean_func)(const RCP<const Basic> &,
+                                                      const RCP<const Basic> &);
+typedef RCP<const Basic> (*vector_arg_func)(const vec_basic &);
+
+// cast overloaded functions below to single_arg, double_arg before they can
+// be used in the map
+static double_arg_boolean_func double_casted_Eq = Eq;
+
+static vector_arg_func vector_casted_add = add;
+static vector_arg_func vector_casted_mul = mul;
+
+static RCP<const Basic> sqr(const RCP<const Basic> &x)
+{
+    return pow(x, integer(2));
+}
+
+static RCP<const Basic> sbml_log(const RCP<const Basic> &b,
+                                 const RCP<const Basic> &x)
+{
+    return log(x, b);
+}
+
+static RCP<const Basic> log10(const RCP<const Basic> &x)
+{
+    return log(x, integer(10));
+}
+
+static RCP<const Basic> root(const RCP<const Basic> &b,
+                             const RCP<const Basic> &x)
+{
+    return pow(x, div(integer(1), b));
+}
+
+static RCP<const Basic> fact(const RCP<const Basic> &x)
+{
+    return gamma(add(x, integer(1)));
+}
+
+static std::map<const std::string,
+                const std::function<RCP<const Basic>(const RCP<const Basic> &)>>
+init_sbml_parser_single_arg_functions()
+{
+    auto functions = init_parser_single_arg_functions();
+    functions.insert({"arcsin", asin});
+    functions.insert({"arccos", acos});
+    functions.insert({"arctan", atan});
+    functions.insert({"arcsec", asec});
+    functions.insert({"arccsc", acsc});
+    functions.insert({"arccot", acot});
+    functions.insert({"arcsinh", asinh});
+    functions.insert({"arccosh", acosh});
+    functions.insert({"arctanh", atanh});
+    functions.insert({"arcsech", asech});
+    functions.insert({"arccoth", acoth});
+    functions.insert({"arccsch", acsch});
```

Comment:
I'm fine with adding these into the main parser.

Reply:
done here: https://github.com/symengine/symengine/pull/1785/commits/8b80e6661a27603487cb28cc1426b8f6a669c27e